File: app/src/main/java/com/absinthe/kage/media/audio/AudioPlayer.java
package com.absinthe.kage.media.audio;

import android.annotation.SuppressLint;
import android.content.Context;
import android.media.session.PlaybackState;
import android.net.wifi.WifiManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.PowerManager;
import android.util.Log;

import com.absinthe.kage.media.LocalMedia;
import com.absinthe.kage.media.PlayList;
import com.absinthe.kage.media.Playback;

import java.util.Observable;

public class AudioPlayer extends Observable implements Playback.Callback {
    public static final String TAG = AudioPlayer.class.getSimpleName();
    public static final String EXTRA_PLAY_MODE = "EXTRA_PLAY_MODE";

    public static final int NOT_REPEATING = 0;
    public static final int REPEAT_ONE = 1;
    public static final int SHUFFLED = 2;
    public static final int REPEAT_ALL = 3;

    public static final int TYPE_LOCAL = 1;
    public static final int TYPE_REMOTE = 2;

    @SuppressLint("StaticFieldLeak")
    private static AudioPlayer sInstance;
    private Context mContext;
    private Handler mHandler = new Handler(Looper.getMainLooper());
    private Playback mPlayback;
    private PlayList mPlaylist;
    private PowerManager.WakeLock mWakeLock;
    private WifiManager.WifiLock mWifiLock;
    private int mBeforePosition = 0;
    private int mPlayMode = REPEAT_ALL;
    private int mPlayState = PlaybackState.STATE_NONE;
    private int mPlayType;

    public static AudioPlayer getInstance(Context context) {
        if (sInstance == null) {
            sInstance = new AudioPlayer(context);
        }
        return sInstance;
    }

    private AudioPlayer(Context context) {
        mContext = context;

        PowerManager pm = (PowerManager) context.getApplicationContext().getSystemService(Context.POWER_SERVICE);
        if (pm != null) {
            mWakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, TAG);
        }
        WifiManager wm = (WifiManager) context.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
        if (wm != null) {
            mWifiLock = wm.createWifiLock(WifiManager.WIFI_MODE_FULL, TAG);
        }
    }

    public void release() {
        mPlayState = PlaybackState.STATE_NONE;
        if (mPlayback != null) {
            mPlayback.stop(false);
        }
        deleteObservers();
        if (mWakeLock != null && mWakeLock.isHeld()) {
            mWakeLock.release();
        }
        mWakeLock = null;
        if (mWifiLock != null && mWifiLock.isHeld()) {
            mWifiLock.release();
        }
        mWifiLock = null;
        sInstance = null;
        mContext = null;
    }

    public void setPlayerType(int type) {
        mPlayType = type;

        if (mPlayback != null) {
            mBeforePosition = mPlayback.getCurrentPosition();
            mPlayback.stop(true);
        }

        if (type == TYPE_LOCAL) {
            mPlayback = new LocalAudioPlayback(mContext);
            mPlayback.setCallback(this);

            if (mPlaylist != null) {
                mPlayback.playMedia(mPlaylist.getCurrentMedia());
            }
        } else if (type == TYPE_REMOTE) {
            mPlayback = new RemoteAudioPlayback();
            mPlayback.setCallback(this);

            if (mPlaylist != null) {
                ((RemoteAudioPlayback) mPlayback).playListMedia(mPlaylist);
            }
        }
    }

    public void playMediaList(PlayList playList) {
        mBeforePosition = 0;
        if (mWakeLock != null && mWakeLock.isHeld()) {
            mWakeLock.acquire(10 * 60 * 1000L /*10 minutes*/);
        }

        if (playList != null) {
            if (mPlaylist == null) {
                mPlaylist = new PlayList();
            }

            mPlaylist.setList(playList.getList(), playList.getCurrentIndex());

            if (mPlayback instanceof RemoteAudioPlayback) {
                ((RemoteAudioPlayback) mPlayback).playListMedia(mPlaylist);
            } else if (mPlayback instanceof LocalAudioPlayback) {
                mPlayback.playMedia(mPlaylist.getCurrentMedia());
            }
        }
    }

    public void play() {
        if (mWakeLock != null && mWakeLock.isHeld()) {
            mWakeLock.acquire(10 * 60 * 1000L /*10 minutes*/);
        }
        mPlayback.play();
    }

    public void pause() {
        mPlayback.pause();
        if (mWakeLock != null && mWakeLock.isHeld()) {
            mWakeLock.release();
        }
    }

    public void playNext() {
        if (mPlaylist != null) {
            playMedia(mPlaylist.getNextMedia(
                    mPlayMode == REPEAT_ONE || mPlayMode == REPEAT_ALL,
                    mPlayMode == SHUFFLED));
        }
    }

    public void playPrevious() {
        if (mPlaylist != null) {
            playMedia(mPlaylist.getPreviousMedia(
                    mPlayMode == REPEAT_ONE || mPlayMode == REPEAT_ALL,
                    mPlayMode == SHUFFLED));
        }
    }

    public void seekTo(int to) {
        mPlayback.seekTo(to);
    }

    public LocalMedia getCurrentMedia() {
        return mPlaylist == null ? null : mPlaylist.getCurrentMedia();
    }

    public int getPlayState() {
        return mPlayState;
    }

    public PlaybackState getPlaybackState() {
        long actions;
        mPlayState = mPlayback.getState();

        if (mPlayState == PlaybackState.STATE_PLAYING) {
            actions = PlaybackState.ACTION_STOP | PlaybackState.ACTION_PAUSE
                    | PlaybackState.ACTION_REWIND | PlaybackState.ACTION_FAST_FORWARD
                    | PlaybackState.ACTION_SEEK_TO;
        } else {
            actions = PlaybackState.ACTION_STOP | PlaybackState.ACTION_PLAY;
        }

        if (hasNext()) {
            actions |= PlaybackState.ACTION_SKIP_TO_NEXT;
        }
        if (hasPre()) {
            actions |= PlaybackState.ACTION_SKIP_TO_PREVIOUS;
        }

        Bundle extras = new Bundle();
        extras.putInt(EXTRA_PLAY_MODE, mPlayMode);

        PlaybackState.Builder builder = new PlaybackState.Builder();
        builder.setActions(actions);
        builder.setState(mPlayState, mPlayback.getCurrentPosition(), 1.0f);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1) {
            builder.setExtras(extras);
        }
        if (mPlaylist != null) {
            builder.setActiveQueueItemId(mPlaylist.getCurrentIndex());
        }
        return builder.build();
    }

    public int getDuration() {
        if (isPlayOrPause()) {
            return mPlayback.getDuration();
        }
        return 0;
    }

    public int getBufferPosition() {
        if (isPlayOrPause()) {
            return mPlayback.getBufferPosition();
        }
        return 0;
    }

    public int getCurrentPosition() {
        if (isPlayOrPause()) {
            return mPlayback.getCurrentPosition();
        }
        return 0;
    }

    public void onCompletion() {
        if (mPlayState == PlaybackState.STATE_PLAYING) {
            playNext();
        } else {
            Log.w(TAG, "mPlayState is not playing");
        }
    }

    public void onPlaybackStateChanged(int state) {
        if (mPlayback != null
                && mPlayState == PlaybackState.STATE_BUFFERING
                && state == PlaybackState.STATE_PLAYING
                && mBeforePosition > 0) {
            Log.d(TAG, "seekTo: " + mBeforePosition);
            mPlayback.seekTo(mBeforePosition);
            mBeforePosition = 0;
        }
        if (mPlayState != state) {
            updateMediaPlayState();
            mPlayState = state;
        }
    }

    public void onError(String error) {
        mPlayState = PlaybackState.STATE_ERROR;
        updateMediaPlayState();
    }

    public void onMediaMetadataChanged(LocalMedia localMedia) {
        updateMediaMetadata(localMedia);
    }

    public int getPlayType() {
        return mPlayType;
    }

    private void updateMediaMetadata(final LocalMedia media) {
        mHandler.post(new Runnable() {
            public void run() {
                synchronized (this) {
                    setChanged();
                    notifyObservers(media);
                }
            }
        });
    }

    private void updateMediaPlayState() {
        mHandler.post(new Runnable() {
            public void run() {
                synchronized (this) {
                    setChanged();
                    notifyObservers(getPlaybackState());
                }
            }
        });
    }

    private boolean hasNext() {
        if (mPlayMode == NOT_REPEATING) {
            return mPlaylist != null && mPlaylist.hasNextMedia();
        }
        return true;
    }

    private boolean hasPre() {
        if (mPlayMode == NOT_REPEATING) {
            return mPlaylist != null && mPlaylist.hasPreviousMedia();
        }
        return true;
    }

    private boolean isPlayOrPause() {
        if (getCurrentMedia() != null) {
            return mPlayState == PlaybackState.STATE_PLAYING || mPlayState == PlaybackState.STATE_PAUSED;
        }
        return false;
    }

    private synchronized void playMedia(LocalMedia media) {
        mBeforePosition = 0;
        mPlayback.playMedia(media);
    }
}

File: app/src/main/java/com/absinthe/kage/media/audio/AudioPlayer.kt
package com.absinthe.kage.media.audio

import android.content.Context
import android.media.session.PlaybackState
import android.os.*
import android.os.PowerManager.WakeLock
import android.util.Log
import com.absinthe.kage.KageApplication
import com.absinthe.kage.device.DeviceManager
import com.absinthe.kage.device.cmd.ResumePlayCommand
import com.absinthe.kage.media.LocalMedia
import com.absinthe.kage.media.PlayList
import com.absinthe.kage.media.Playback
import java.util.*

object AudioPlayer: Observable(), Playback.Callback {

    const val TAG = "AudioPlayer"
    const val TYPE_LOCAL = 1
    const val TYPE_REMOTE = 2

    private const val EXTRA_PLAY_MODE = "EXTRA_PLAY_MODE"
    private const val NOT_REPEATING = 0
    private const val REPEAT_ONE = 1
    private const val SHUFFLED = 2
    private const val REPEAT_ALL = 3

    private var mPlayback: Playback? = null
    private var mPlaylist: PlayList? = null
    private var mWakeLock: WakeLock
    private var mPosition = 0

    private val mHandler = Handler(Looper.getMainLooper())
    private val mPlayMode = NOT_REPEATING

    private var playType = TYPE_LOCAL
    var playState = PlaybackState.STATE_NONE
        private set

    init {
        val pm = KageApplication.sContext.getSystemService(Context.POWER_SERVICE) as PowerManager
        mWakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "$TAG:WakeLock")
    }

    fun release() {
        playState = PlaybackState.STATE_NONE
        if (mPlayback != null) {
            mPlayback!!.stop(false)
        }
        deleteObservers()
        if (mWakeLock.isHeld) {
            mWakeLock.release()
        }
    }

    fun setPlayerType(type: Int) {
        playType = type
        if (mPlayback != null) {
            mPosition = mPlayback!!.currentPosition
            mPlayback!!.stop(true)
        }
        if (type == TYPE_LOCAL) {
            mPlayback = LocalAudioPlayback(KageApplication.sContext)
            (mPlayback as LocalAudioPlayback).setCallback(this)
        } else if (type == TYPE_REMOTE) {
            mPlayback = RemoteAudioPlayback()
            (mPlayback as RemoteAudioPlayback).setCallback(this)
        }
        if (mPlayback != null && mPlaylist != null) {
            mPlayback!!.playMedia(mPlaylist!!.currentMedia)
        }
    }

    @Synchronized
    fun playMedia(media: LocalMedia) {
        mPosition = 0
        mPlayback!!.playMedia(media)
        if (mPlaylist == null) {
            mPlaylist = PlayList()
        }
        mPlaylist!!.addMedia(media)
        mPlaylist!!.currentIndex = 0
    }

    fun playMediaList(playList: PlayList?) {
        mPosition = 0
        if (mWakeLock.isHeld) {
            mWakeLock.acquire(10 * 60 * 1000L /*10 minutes*/)
        }
        if (playList != null) {
            if (mPlaylist == null) {
                mPlaylist = PlayList()
            }
            mPlaylist!!.setList(playList.list, playList.currentIndex)
            if (mPlayback is RemoteAudioPlayback) {
                (mPlayback as RemoteAudioPlayback).playListMedia(mPlaylist!!)
            } else if (mPlayback is LocalAudioPlayback) {
                (mPlayback as LocalAudioPlayback).playMedia(mPlaylist!!.currentMedia)
            }
        }
    }

    fun play() {
        if (mWakeLock.isHeld) {
            mWakeLock.acquire(10 * 60 * 1000L /*10 minutes*/)
        }
        if (mPlayback != null) {
            mPlayback!!.play()
        }
    }

    fun pause() {
        if (mPlayback != null) {
            mPlayback!!.pause()
        }
        if (mWakeLock.isHeld) {
            mWakeLock.release()
        }
    }

    fun resumePlay() {
        val deviceManager = DeviceManager.Singleton.INSTANCE.instance
        if (deviceManager.isConnected) {
            deviceManager.sendCommandToCurrentDevice(ResumePlayCommand())
        }
    }

    fun playNext() {
        if (mPlaylist != null) {
            playMedia(mPlaylist!!.getNextMedia(
                    mPlayMode == REPEAT_ONE || mPlayMode == REPEAT_ALL,
                    mPlayMode == SHUFFLED))
        }
    }

    fun playPrevious() {
        if (mPlaylist != null) {
            playMedia(mPlaylist!!.getPreviousMedia(
                    mPlayMode == REPEAT_ONE || mPlayMode == REPEAT_ALL,
                    mPlayMode == SHUFFLED))
        }
    }

    fun seekTo(to: Int) {
        if (mPlayback != null) {
            mPlayback!!.seekTo(to)
        }
    }

    val currentMedia: LocalMedia?
        get() = if (mPlaylist == null) null else mPlaylist!!.currentMedia

    val playbackState: PlaybackState
        get() {
            var actions: Long
            playState = mPlayback!!.state
            actions = if (playState == PlaybackState.STATE_PLAYING) {
                (PlaybackState.ACTION_STOP or PlaybackState.ACTION_PAUSE
                        or PlaybackState.ACTION_REWIND or PlaybackState.ACTION_FAST_FORWARD
                        or PlaybackState.ACTION_SEEK_TO)
            } else {
                PlaybackState.ACTION_STOP or PlaybackState.ACTION_PLAY
            }

            if (hasNext()) {
                actions = actions or PlaybackState.ACTION_SKIP_TO_NEXT
            }
            if (hasPre()) {
                actions = actions or PlaybackState.ACTION_SKIP_TO_PREVIOUS
            }

            val extras = Bundle()
            extras.putInt(EXTRA_PLAY_MODE, mPlayMode)
            val builder = PlaybackState.Builder()
            builder.setActions(actions)
            builder.setState(playState, mPlayback!!.currentPosition.toLong(), 1.0f)

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1) {
                builder.setExtras(extras)
            }
            if (mPlaylist != null) {
                builder.setActiveQueueItemId(mPlaylist!!.currentIndex.toLong())
            }

            return builder.build()
        }

    val duration: Int
        get() = if (isPlayOrPause) {
            mPlayback!!.duration
        } else 0

    val bufferPosition: Int
        get() = if (isPlayOrPause) {
            mPlayback!!.bufferPosition
        } else 0

    val currentPosition: Int
        get() = if (isPlayOrPause) {
            mPlayback!!.currentPosition
        } else 0

    override fun onCompletion() {
        if (playState == PlaybackState.STATE_PLAYING) {
            playNext()
        } else {
            Log.w(TAG, "PlayState is not playing")
        }
    }

    override fun onPlaybackStateChanged(state: Int) {
        if (mPlayback != null && playState == PlaybackState.STATE_BUFFERING && state == PlaybackState.STATE_PLAYING && mPosition > 0) {
            Log.d(TAG, "Seek to: $mPosition")
            mPlayback!!.seekTo(mPosition)
            mPosition = 0
        }
        if (playState != state) {
            updateMediaPlayState()
            playState = state
        }
    }

    override fun onError(error: String) {
        playState = PlaybackState.STATE_ERROR
        updateMediaPlayState()
    }

    override fun onMediaMetadataChanged(localMedia: LocalMedia) {
        updateMediaMetadata(localMedia)
    }

    private fun updateMediaMetadata(media: LocalMedia) {
        mHandler.post(object : Runnable {
            override fun run() {
                synchronized(this) {
                    setChanged()
                    notifyObservers(media)
                }
            }
        })
    }

    private fun updateMediaPlayState() {
        mHandler.post(object : Runnable {
            override fun run() {
                synchronized(this) {
                    setChanged()
                    notifyObservers(playbackState)
                }
            }
        })
    }

    private operator fun hasNext(): Boolean {
        return mPlaylist != null && mPlaylist!!.hasNextMedia()
    }

    private fun hasPre(): Boolean {
        return mPlaylist != null && mPlaylist!!.hasPreviousMedia()
    }

    private val isPlayOrPause: Boolean
        get() = if (currentMedia != null) {
            playState == PlaybackState.STATE_PLAYING || playState == PlaybackState.STATE_PAUSED
        } else false
}